### ISSUE TITLE
Make derivative open-dispatch by converting its implementation to protocols

### ISSUE THREAD

**Issue by sritchie** [sicmutils/sicmutils#92]
I don't know the implementation well yet, but I wanted to flag this section as ripe for protocol conversion. https://github.com/littleredcomputer/sicmutils/blob/master/src/sicmutils/calculus/derivative.cljc#L218

the generic arithmetic is totally open now. I think a nice test would be - can someone implement Quaternions in their code and have everything just work, given some set of protocol and multimethod implementations?

Making a ticket to get your thoughts, @littleredcomputer , and to track notes on how we might make this happen.

**Comment by littleredcomputer** [comment]
This is interesting and as I look at the code I realize, with nostalgia, that I used `letfn` because GJS used nested define and my plan was not to disturb the structure of things too much because I was going to need to parallel-compare implementations while I figured out how it all worked. I agree this should become open-dispatch. On differentiation, or a `fmap`-like operation, that shifts a function application from a container to its elements?
